### ISSUE TITLE
Labels push HKDF-Expand-Label into two hash blocks

### ISSUE THREAD

**Issue by ekr** [tlswg/tls13-spec#964]
I notice that some of the application traffic secret labels are long enough to push us into a second hash block. I don't think this is a big deal, but I thought I would mention it because now is the time to change.

@adl @martinthomson @davidben

**Comment by ekr** [comment]
I think someone mentioned this in ORD....

**Comment by davidben** [comment]
*shrug* I also don't think it's a big deal, but if someone feels strongly, I also don't object to shrinking them.

(Which labels?)

**Comment by martinthomson** [comment]
I also don't care, but I know that @adl was sad about the number of times we cranked the handle on the hash function (and that was before we added many).  I wouldn't object to someone trimming the strings a little if it came to that.

**Comment by davidben** [comment]
<unhelpful>We could pump all our labels into the hash function. That would make them shorter!</unhelpful> :-)

**Comment by davegarrett** [comment]
Chopping off the " secret" from the end of everything would be a quick easy shortening, if that's enough.

**Comment by ekr** [comment]
I'd have to make a list, but I think minimally client application traffic secret is going to be a problem because it has to be concatenated to the hash context and some framing info.

**Comment by ilaril** [comment]
I calculated that the maximum label length for derive-secret is 18 bytes if one is not to blow SHA-256 block:

* 2 bytes: Output length (always 32)
* 1 byte: Length of label + 9
* 9 bytes: "TLS 1.3, "
* 18 bytes: Label
* 1 byte: Length of context (always 32)
* 32 bytes: hash
* 1 byte: HKDF trailer (always 1)

Total: 64 bytes (block is 64 bytes).

So just chopping "secret" is not enough.

SHA-384 uses 80 bytes of its 128 byte block at 18 byte label, so it fits too.

(18 bytes would also be small enough not to make most other hashes not blow extra block. The most notable exception would be 512-bit SHA-3 (not to be confused with 512-bit SHAKE256), but that hash function is not very usable due to slowness anyway).

**Comment by ekr** [comment]
Here is a revised set of labels of which the longest is 18.

external binder # was external psk binder key
resumption binder # was resumption psk binder key
client e. traffic # was client early traffic
e. exporter master # was early exporter master secret
client hs traffic # was client handshake traffic secret
server hs traffic # was server handshake traffic secret
client app traffic # was client application traffic secret
server app traffic # was server application traffic secret
exporter master # was exporter master secret
resumption # was resumption master secret
key # was key
iv # was iv
finished # was finished
app traffic upd.  # was application traffic secret
exporter # was exporter

**Comment by davidben** [comment]
In the category of things that don't matter:

"app traffic upd." could be "key update" which avoids abbreviation and matches the name of KeyUpdate. It loses the "app" distinguisher, but even if 0-RTT or HS data were to have a KeyUpdate, not having a distinguisher is fine for the reason keys and IVs have short labels.

"e. exporter master" could be "early exporter" to avoid the abbreviation and also match "exporter". That said, avoiding the abbreviation means inconsistency with "client e. traffic". "client 0rtt traffic" and "0rtt exporter" do fit, but the spec tends to call it "early".

In other words: I see you have proposed a mauvy shade of pinky russet. I don't care, but have you considered a slightly greener color? :-)

**Comment by ekr** [comment]
I agree with you about "app traffic upd." will change.

I did spend a while messing around with the earlys and masters but ran into the same problem you did, and thus ended up with the result you see here.

**Comment by ekr** [comment]
Thus, we end up with a slightly bluer color :)

**Comment by ekr** [comment]
Revised:

external binder # was external psk binder key
resumption binder # was resumption psk binder key
client e. traffic # was client early traffic
e. exporter master # was early exporter master secret
client hs traffic # was client handshake traffic secret
server hs traffic # was server handshake traffic secret
client app traffic # was client application traffic secret
server app traffic # was server application traffic secret
exporter master # was exporter master secret
resumption # was resumption master secret
key # was key
iv # was iv
finished # was finished
traffic key update  # was application traffic secret
exporter # was exporter

**Comment by enygren** [comment]
Is there value in namespacing the labels (eg, "tls.key" and "tls.tfk.app.c"?), even if it means shortening some?  This might help avoid unfortunate collisions with extensions down the road?
And/or otherwise adding an IANA registry for the labels?
Or (shudder) using OIDs for labels?
(What the label string values presumably don't matter as much as that they have a good mapping to the specification and are clearly/unambiguously unique.)

**Comment by richsalz** [comment]
An FCFS label registry seems to make sense.

**Comment by ekr** [comment]
Revised based on ML discussion

16 tls13 ext binder    #  was external psk binder key
16 tls13 res binder    #  was resumption psk binder key
17 tls13 c e traffic    #  was client early traffic secret
18 tls13 e exp master    #  was early exporter master secret
18 tls13 c hs traffic    #  was client handshake traffic secret
18 tls13 s hs traffic    #  was server handshake traffic secret
18 tls13 c ap traffic    #  was client application traffic secret
18 tls13 s ap traffic    #  was server application traffic secret
16 tls13 exp master    #  was exporter master secret
16 tls13 res master    #  was resumption master secret
9 tls13 key    #  was key
8 tls13 iv    #  was iv
14 tls13 finished    #  was finished
17 tls13 traffic upd    #  was application traffic secret
14 tls13 exporter    #  was exporter
13 tls13 derived    #  was derived

**Comment by ad-l** [comment]
This does save some unnecessary compressions so I am in favour of this change.
My only comment is that the "tls13" prefix seems unnecessary given the hkdfLabel encoding:
```C
       struct {
           uint16 length = Length;
           opaque label<10..255> = "TLS 1.3, " + Label;
           opaque hash_value<0..255> = HashValue;
       } HkdfLabel;
```
I do not believe that HKDF-Expand is used directly for any of the above labels. Since HKDF-Expand-Label already injects the "TLS 1.3, " prefix I see no point in having it in the label too.

**Comment by ekr** [comment]
The intent is to have that inject "tls13 " and as you say, no redundant labels.
